engine/src/main/java/org/hibernate/validator/internal/constraintvalidators/hv/ISBNValidator.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * Hibernate Validator, declare and validate application constraints
+ *
+ * License: Apache License, Version 2.0
+ * See the license.txt file in the root directory or <http://www.apache.org/licenses/LICENSE-2.0>.
+ */
+package org.hibernate.validator.internal.constraintvalidators.hv;
+
+import java.util.function.Function;
+
+import javax.validation.ConstraintValidator;
+import javax.validation.ConstraintValidatorContext;
+
+import org.hibernate.validator.constraints.ISBN;
+
+/**
+ * Checks that a given character sequence (e.g. string) is a valid ISBN.
+ *
+ * @author Marko Bekhta
+ */
+public class ISBNValidator implements ConstraintValidator<ISBN, CharSequence> {
+
+	private int length;
+
+	private Function<String, Boolean> checkChecksumFunction;
+
+	@Override
+	public void initialize(ISBN constraintAnnotation) {
+		switch ( constraintAnnotation.type() ) {
+			case ISBN10:
+				length = 10;
+				checkChecksumFunction = this::checkChecksumISBN10;
+				break;
+			case ISBN13:
+				length = 13;
+				checkChecksumFunction = this::checkChecksumISBN13;
+				break;
+		}
+	}
+
+	@Override
+	public boolean isValid(CharSequence isbn, ConstraintValidatorContext context) {
+		if ( isbn == null ) {
+			return true;
+		}
+
+		// Replace all non-digit chars
+		String digits = isbn.toString().replaceAll( "[^\\dX]", "" );
```

Comment:
I think this would be better with a precompiled pattern. Keep in mind that's what `replaceAll()` uses under the hood so the pattern ends up being built every time.

Reply:
By the way, the comment is not accurate as you also keep the `X` character.